### PR TITLE
drivers: hwinfo: add driver for TI SimpleLink CC13xx/CC26xx series

### PR DESCRIPTION
**Pull request by pepe2k**
Introduce hwinfo driver for **Texas Instruments SimpleLink CC13xx/CC26xx** SOC series, with support for retrieving device's ID and cause of reset.

By default, the pre-programmed, 8-bytes length IEEE MAC address is used for device's ID, with option (`HWINFO_CC13XX_CC26XX_USE_BLE_MAC`) for use 6-bytes length BLE MAC instead. By default, read-only values stored in `FCFG` (Factory Configuration) are used but additional Kconfig option (`HWINFO_CC13XX_CC26XX_ALWAYS_USE_FACTORY_DEFAULT`) is also provided, which allows to use value from writable `CCFG` (Customer Configuration) area. In all cases, device ID value is provided in big-endian which matches what TI's UniFlash tool returns.

For reset cause, `SysCtrlResetSourceGet()` function from TI's driver library is used and its value is mapped to Zephyr's implementation. Importantly, only root cause of reset is available on CC13xx/CC26xx platform as described in MCU Technical Reference Manual (`RESET_SRC` field in `RESETCTL` register):

> Shows the root cause of the last system reset. More than the reported reset source can have been active during the last system reset but only the root cause is reported.

### CODE REVIEW

File: drivers/hwinfo/hwinfo_cc13xx_cc26xx.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_cc13xx_cc26xx.c:37]: Can you add comment on which #if #else path is closing?
Otherwise looks good to me.
Thanks!

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/hwinfo/hwinfo_cc13xx_cc26xx.c
Comment (pepe2k) [reply drivers/hwinfo/hwinfo_cc13xx_cc26xx.c]: Done.
Was wondering whether indentations of preprocessor statements are accepted/allowed?
E.g. something like:

```
#ifdef ...
#  ifdef ...
#  else
#  endif
#endif
```

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_cc13xx_cc26xx.c]: Just the comments as far as I know.

**Review by alexanderwachter — DISMISSED** [review]

**Comment by alexanderwachter** [comment]
@carlescufi @henrikbrixandersen 
Can we have a second review here please.

**Review by cfriedt — DISMISSED** [review]
LGTM, although personally, I am not a big fan of changing the whole CMakeLists.txt file to align arguments.

**Review by vaishnavachath — DISMISSED** [review]

File: drivers/hwinfo/hwinfo_cc13xx_cc26xx.c
Comment (fgrandel) [inline drivers/hwinfo/hwinfo_cc13xx_cc26xx.c:27]: Unless I overlook some technical reason why we need defines, I believe there are reasons to prefer the following pattern:

```suggestion
if (IS_ENABLED(CONFIG_HWINFO_CC13XX_CC26XX_USE_BLE_MAC) || sys_read32(CCFG_BASE + CCFG_O_IEEE_BLE_0) == 0xFFFFFFFF || sys_read32(CCFG_BASE + CCFG_O_IEEE_BLE_1) == 0xFFFFFFFF) {
...
}
```

* easier to read and maintain
* ensures that all code is compiled independently of compile time config options (avoids bit rot).
* In your specific case it also removes redundant code

The compiler will optimize unused code away so that the end result is not wasting more flash - at least if you put the check on the constant value first in the expression. You can easily check this in the resulting assembly code if you look at the result through some debugger frontend.

**Comment by pepe2k** [comment]
@cfriedt 

> LGTM, although personally, I am not a big fan of changing the whole CMakeLists.txt file to align arguments.

I wasn't sure about that and I'm fine with dropping alignment related change here with just single-line addition left.
What do you suggest?

Thanks!

File: drivers/hwinfo/Kconfig
Comment (fgrandel) [inline drivers/hwinfo/Kconfig:39]: You may (optionally) consider using a choice value for this with two options: BLE vs. 15.4. IMO that would be easier to understand for the end user. To me it's not immediately obvious why BLE and/or 15.4 MAC should be preferred.

File: drivers/hwinfo/CMakeLists.txt
Comment (fgrandel) [inline drivers/hwinfo/CMakeLists.txt:11]: As @cfriedt also suggested, you should only introduce your own line and leave the other lines alone. Otherwise we loose history on those. We generally prefer to live with a little less nicely formatted code instead.

Not a blocker, though, IMO.

**Review by ghost — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. Looks good! :-)

Just one minor thing - see comment below re IS_ENABLED, the rest are optional nitpicks. Maybe I'm overlooking something, but if it works I think you'll also like that version more. Otherwise let me know.

File: drivers/hwinfo/hwinfo_cc13xx_cc26xx.c
Comment (pepe2k) [reply drivers/hwinfo/hwinfo_cc13xx_cc26xx.c]: I switched to `IS_ENABLED()` but in a slightly different way ('always use factory defaults' support).
Thanks for the suggestion!

File: drivers/hwinfo/Kconfig
Comment (pepe2k) [reply drivers/hwinfo/Kconfig:51]: I switched to `menuconfig` for the whole driver Kconfig symbol and `choice` for BLE/IEEE MAC selection, thanks!

File: drivers/hwinfo/CMakeLists.txt
Comment (pepe2k) [reply drivers/hwinfo/CMakeLists.txt:11]: Done, just a single-line change here.

**Review by ghost — APPROVED** [review]
Nice! Thanks a lot for these changes.

**Review by vaishnavachath — APPROVED** [review]

**Review by alexanderwachter — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by kartben** [comment]
@pepe2k  Sorry that I missed the opportunity to comment earlier, but I think you will want to update the yaml files for all the relevant boards to add `hwinfo` to the `supported` list.

**Comment by cfriedt** [comment]
> @pepe2k  Sorry that I missed the opportunity to comment earlier, but I think you will want to update the yaml files for all the relevant boards to add `hwinfo` to the `supported` list.

Alternatively, @vaishnavachath

**Comment by pepe2k** [comment]
@kartben thank you for your suggestion, I have created relevant PR: https://github.com/zephyrproject-rtos/zephyr/pull/63083, btw I'm not sure whether information about hwinfo support should go also into documentation, under "supported features". Seems like there is some inconsistency regarding that.

**Pull request merged**
